### PR TITLE
Peers RPC for Bucket Notification and Listeners

### PR DESCRIPTION
**Pull request by donatello**

### CODE REVIEW

**Comment by donatello** [comment]
This PR is intended to resolve one part of https://github.com/minio/minio/issues/2772:
- Bucket notification configuration is activated on all minio servers via internal RPC calls during a PutBucketNotification API call - this means all servers start sending out events to external services (like redis, postgres, elasticsearch, amqp, nats, etc).
- Info about listeners is also sent to all minio servers via RPC calls during a ListenBucketNotification API call - all servers then start sending out events to the minio server connected to the listening client.

File: cmd/event-notifier.go [hunk 1]
```diff
@@ -29,14 +30,52 @@ import (
...
+}
 
-	// Collection of 'bucket' and notification config.
-	notificationConfigs map[string]*notificationConfig
-	snsTargets          map[string][]chan []NotificationEvent
-	queueTargets        map[string]*logrus.Logger
+type internalNotifier struct {
+	// per-bucket listener configuration. This is updated
+	// when listeners connect or disconnect.
+	//listenerConfigs map[listenerID]*listenerConfig
```
Comment (donatello) [inline cmd/event-notifier.go]: Fix comment.

File: cmd/event-notifier.go [hunk 2]
```diff
@@ -202,43 +247,58 @@ func eventNotify(event eventData) {
...
-			if targetLog != nil {
-				targetLog.WithFields(logrus.Fields{
-					"Key":       path.Join(event.Bucket, objectName),
-					"EventType": eventType,
-					"Records":   notificationEvent,
-				}).Info()
+	nConfig := globalEventNotifier.GetBucketNotificationConfig(event.Bucket)
+	lCfgs := globalEventNotifier.GetBucketListenerConfig(event.Bucket)
+
+	if nConfig != nil {
```
Comment (donatello) [inline cmd/event-notifier.go]: Add comment for readability.

File: cmd/event-notifier.go [hunk 3]
```diff
@@ -267,36 +328,112 @@ func loadNotificationConfig(bucket string, objAPI ObjectLayer) (*notificationCon
...
+		}
+		errorIf(err, "Unable to load bucket-listeners for bucket %s", bucket)
+		// Returns error for other errors.
+		return nil, err
+	}
+
+	// Unmarshal notification bytes.
+	var lCfg []listenerConfig
+	lConfigBytes := buffer.Bytes()
+	if err = json.Unmarshal(lConfigBytes, &lCfg); err != nil {
```
Comment (donatello) [inline cmd/event-notifier.go:390]: `gjson`?

File: cmd/event-notifier.go [hunk 4]
```diff
@@ -267,36 +328,112 @@ func loadNotificationConfig(bucket string, objAPI ObjectLayer) (*notificationCon
...
+		errorIf(err, "Listener config - unmarshal error!")
```
Comment (donatello) [inline cmd/event-notifier.go]: `Unable to marshal...` - error message consistency

File: cmd/s3-peer-client.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	"path"
+	"time"
+)
+
+type s3Peers struct {
+	// A map of peer server names to RPC client connections
+	rpcClient map[string]*AuthRPCClient
+
+	// slice of all peers
+	addr []string
```
Comment (donatello) [inline cmd/s3-peer-client.go]: `addrs` or `peers`

File: cmd/s3-peer-client.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+// returns the network addresses of all Minio servers in the cluster
+func getAllPeers(disks []string) []string {
+	res := []string{}
+	for _, disk := range disks {
+		netAddr, _, err := splitNetPath(disk)
+		if err != nil || netAddr == "" {
+			// If this was a problem error, it would be
+			// found earlier in the startup
```
Comment (donatello) [inline cmd/s3-peer-client.go]: log the error cases.

File: cmd/utils.go [hunk 7]
```diff
@@ -80,19 +80,66 @@ func splitNetPath(networkPath string) (netAddr, netPath string, err error) {
...
-	} else if !filepath.IsAbs(networkParts[1]) {
+	case !filepath.IsAbs(networkParts[1]):
 		return "", "", &net.AddrError{Err: "Network path should be absolute", Addr: networkPath}
 	}
 	return networkParts[0], networkParts[1], nil
 }
 
+// returns the local server address by parsing the disk paths from the
+// command line - intended to initialize `globalMinioAddr`
+func getLocalAddress(serverAddr string, disks []string) string {
```
Comment (donatello) [inline cmd/utils.go]: Reuse `localNode` an existing function in `control-router.go`

File: cmd/s3-peer-client.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+type s3Peers struct {
+	// A map of peer server names to RPC client connections
+	rpcClient map[string]*AuthRPCClient
+
+	// slice of all peers
+	addr []string
+}
+
+func initGlobalS3Peers(disks []string) {
```
Comment (donatello) [inline cmd/s3-peer-client.go:36]: Make list of peers unique.

File: cmd/event-notifier.go (hunk 2 above)
Comment (donatello) [reply cmd/event-notifier.go]: Refactored this part to use separate functions for listen and notifications.

File: cmd/event-notifier.go (hunk 3 above)
Comment (donatello) [reply cmd/event-notifier.go:390]: @harshavardhana `gjson` does not seem to be good to unmarshall arrays. It returns `[]gjson.Result` and each of these `Result`s need further processing to load into `listenerConfig`. I think its better to stick with `encoding/json` for this case.

Comment (harshavardhana) [reply cmd/event-notifier.go:390]: Understood.. we will be fine..

**Comment by harshavardhana** [comment]
ping @donatello - the tests have failed on linux looks like getLocalAddress has some issue.

**Comment by donatello** [comment]
Yes, will tackle this shortly.

File: cmd/utils_test.go [hunk 9]
```diff
@@ -174,3 +175,65 @@ func TestMaxPartID(t *testing.T) {
 		}
 	}
 }
+
+// Tests fetch local address.
+func TestLocalAddress(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		return
+	}
+	// need to set this to avoid stale values from other tests.
```
Comment (donatello) [inline cmd/utils_test.go:184]: @harshavardhana Test fixed!

**Pull request merged**